### PR TITLE
[WIP] add workloadEntry into API

### PR DESCRIPTION
**Pull request by irisdingbj**
@smawson Please take a look at this change to see whether this is what we want. Thanks.

For
https://docs.google.com/document/d/1VDV4-prbMbzW06hslfp12sn7V6oUvnIAst6mT6FtSWc/edit#heading=h.xw1gqgyqs5b

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
What is it diff from ServiceEntry? Any design link?

**Comment by irisdingbj** [comment]
> What is it diff from ServiceEntry? Any design link?

https://docs.google.com/document/d/1VDV4-prbMbzW06hslfp12sn7V6oUvnIAst6mT6FtSWc/edit#heading=h.uakvommwzrq2

ServiceEntry can select workloadEntry.

File: networking/v1alpha3/workload_entry.proto [hunk 1]
```diff
@@ -0,0 +1,102 @@
+// Copyright 2018 Istio Authors
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]: 2020

File: networking/v1alpha3/workload_entry.proto [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+syntax = "proto3";
+
+import "google/api/field_behavior.proto";
+import "networking/v1alpha3/gateway.proto";
+
+
+// $schema: istio.networking.v1alpha3.WorkloadEntry
+// $title: Workload Entry
+// $description: Configuration affecting workload  entry.
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]:
```suggestion
// $description: Configuration of a workload entry.
```

File: networking/v1alpha3/workload_entry.proto [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+// `WorkloadEntry` enables adding additional workloads into Istio's internal
+// service registry, so that auto-discovered services in the mesh can
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]:
```suggestion
// service registry, so that services in the mesh can
```

File: networking/v1alpha3/workload_entry.proto [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+// access/route to these manually specified workload. A workload entry
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]:
```suggestion
// route to these workloads. A workload entry
```

File: networking/v1alpha3/workload_entry.proto [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+// describes the properties of a workload.
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]:
```suggestion
// describes the properties of a workload such as its identity and endpoints.
```

File: networking/v1alpha3/workload_entry.proto [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+//
+// The following example declares a  workloadEntry running on both 1.2.3.4 and 1.2.3.5.
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]:
```suggestion
// The following example declares a workloadEntry running on two IP addresses: 1.2.3.4 and 1.2.3.5.
```

File: networking/v1alpha3/workload_entry.proto [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+import "google/api/field_behavior.proto";
+import "networking/v1alpha3/gateway.proto";
+
+
+// $schema: istio.networking.v1alpha3.WorkloadEntry
+// $title: Workload Entry
+// $description: Configuration affecting workload  entry.
+
+// `WorkloadEntry` enables adding additional workloads into Istio's internal
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]:
```suggestion
// `WorkloadEntry` enables adding additional workloads such as VMs into Istio's internal
```

File: networking/v1alpha3/workload_entry.proto [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+// describes the properties of a workload.
+//
+// The following example declares a  workloadEntry running on both 1.2.3.4 and 1.2.3.5.
+// expose ports on 27018
+//
+// ```yaml
+// apiVersion: networking.istio.io/v1alpha3
+// kind: WorkloadEntry
+// metadata:
+//   name: myvm
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]: Should stick it in a namespace just as best practice?

File: networking/v1alpha3/workload_entry.proto [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+//   - number: 27018
+//     name: mongodb
+//     protocol: MONGO
+//   endpoints:
+//     - addresses:
+//       - 1.2.3.4
+//       - 1.2.3.5
+// ```
+//
+// The following example declares a workloadEntry running on  1.2.3.4
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]:
```suggestion
// The following example declares a WorkloadEntry running at 1.2.3.4
```

File: networking/v1alpha3/workload_entry.proto [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+// and expose ports on 27018 and 8888
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]:
```suggestion
// which exposes ports 27018 and 8888.
```

File: networking/v1alpha3/workload_entry.proto [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+// $title: Workload Entry
+// $description: Configuration affecting workload  entry.
+
+// `WorkloadEntry` enables adding additional workloads into Istio's internal
+// service registry, so that auto-discovered services in the mesh can
+// access/route to these manually specified workload. A workload entry
+// describes the properties of a workload.
+//
+// The following example declares a  workloadEntry running on both 1.2.3.4 and 1.2.3.5.
+// expose ports on 27018
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]:
```suggestion
// Both endpoints expose port 27018.
```

File: networking/v1alpha3/workload_entry.proto [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+//
+
+package istio.networking.v1alpha3;
+
+option go_package = "istio.io/api/networking/v1alpha3";
+
+
+message WorkloadEntry {
+
+  string serviceAccountName = 1 [(google.api.field_behavior) = REQUIRED];
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]: Add comment describing the meaning of this field.

File: networking/v1alpha3/workload_entry.proto [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+  // One or more labels associated with the workloadEntry.
+  map<string, string> labels = 2;
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]: Shouldn't this be part of metadata? I don't think we need additional labels.

File: networking/v1alpha3/workload_entry.proto [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+  repeated Port ports = 3;
+
+  // One or more endpoints associated with the workloadEntry.
+  repeated Endpoint endpoints = 4 [(google.api.field_behavior) = REQUIRED];
+
+  };
+
+  message Endpoint {
+
+    repeated string addresses = 1 [(google.api.field_behavior) = REQUIRED];
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]: Describe what types of addresses are supported.

File: networking/v1alpha3/workload_entry.proto [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+//
+// ```yaml
+// apiVersion: networking.istio.io/v1alpha3
+// kind: WorkloadEntry
+// metadata:
+//   name: myvm
+//   labels:
+//      app: myapp
+//      version: v1
+// spec:
```
Comment (howardjohn) [inline networking/v1alpha3/workload_entry.proto]: Do we need the topology here as well? Otherwise we will have no way to extract this information.

File: networking/v1alpha3/workload_entry.proto (hunk 14 above)
Comment (howardjohn) [inline networking/v1alpha3/workload_entry.proto]: What types of address are supported? IP only or DNS as well? We will need to document how this should be used

Comment (howardjohn) [reply networking/v1alpha3/workload_entry.proto]: In particular the difference between
```
endpoints:
- addresses: [a, b, c]
```
and 

```
endpoints:
- addresses: [a]
- addresses: [b]
- addresses: [c]
```

Which should be that the first is one instance with multiple addresses (dual stack, etc), the second being 3 different instances

File: networking/v1alpha3/workload_entry.proto (hunk 13 above)
Comment (howardjohn) [inline networking/v1alpha3/workload_entry.proto]: Do we want labels defined here, given its already in k8s object meta?

File: networking/v1alpha3/workload_entry.proto (hunk 1 above)
Comment (irisdingbj) [reply networking/v1alpha3/workload_entry.proto]: fixed

File: networking/v1alpha3/workload_entry.proto (hunk 2 above)
Comment (irisdingbj) [reply networking/v1alpha3/workload_entry.proto]: fixed

File: networking/v1alpha3/workload_entry.proto (hunk 4 above)
Comment (irisdingbj) [reply networking/v1alpha3/workload_entry.proto]: fixed

File: networking/v1alpha3/workload_entry.proto (hunk 5 above)
Comment (irisdingbj) [reply networking/v1alpha3/workload_entry.proto]: fixed

File: networking/v1alpha3/workload_entry.proto (hunk 11 above)
Comment (irisdingbj) [reply networking/v1alpha3/workload_entry.proto]: fixed

File: networking/v1alpha3/workload_entry.proto (hunk 13 above)
Comment (irisdingbj) [reply networking/v1alpha3/workload_entry.proto]: removed

Comment (irisdingbj) [reply networking/v1alpha3/workload_entry.proto]: removed

File: networking/v1alpha3/workload_entry.proto (hunk 14 above)
Comment (irisdingbj) [reply networking/v1alpha3/workload_entry.proto]: added explanation we support both formats, and also we support both DNS and static IP

File: networking/v1alpha3/workload_entry.proto (hunk 12 above)
Comment (irisdingbj) [reply networking/v1alpha3/workload_entry.proto]: fixed

File: networking/v1alpha3/workload_entry.proto (hunk 14 above)
Comment (irisdingbj) [reply networking/v1alpha3/workload_entry.proto]: fixed

File: networking/v1alpha3/workload_entry.proto (hunk 15 above)
Comment (irisdingbj) [reply networking/v1alpha3/workload_entry.proto]: added  topoloty field like the one from k8s.endpoint

**Comment by irisdingbj** [comment]
@smawson @howardjohn  Thanks for the review comments. PTAL.

File: networking/v1alpha3/workload_entry.proto (hunk 3 above)
Comment (smawson) [reply networking/v1alpha3/workload_entry.proto]: Wasn't changed, what do you mean by "auto-discovered" here?

File: networking/v1alpha3/workload_entry.proto [hunk 16]
```diff
@@ -20,15 +20,15 @@ import "networking/v1alpha3/gateway.proto";
...
 // $schema: istio.networking.v1alpha3.WorkloadEntry
 // $title: Workload Entry
-// $description: Configuration affecting workload  entry.
+// $description: Configuration of a workload entry.
 
 // `WorkloadEntry` enables adding additional workloads into Istio's internal
 // service registry, so that auto-discovered services in the mesh can
-// access/route to these manually specified workload. A workload entry
-// describes the properties of a workload.
+// route to these workloads. A workload entry. A workload entry
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]:
```suggestion
// route to these workloads. A workload entry
```

File: networking/v1alpha3/workload_entry.proto (hunk 10 above)
Comment (smawson) [reply networking/v1alpha3/workload_entry.proto]: Not fixed

File: networking/v1alpha3/workload_entry.proto [hunk 17]
```diff
@@ -75,28 +75,112 @@ import "networking/v1alpha3/gateway.proto";
 //       - 1.2.3.4
 // ```
 //
+// The following example declares a workloadEntry running with dual IP stack on 1.2.3.4 and 1.2.3.5
+// and expose ports on 27018 and 8888
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]:
```suggestion
// which exposes ports 27018 and 8888.
```

File: networking/v1alpha3/workload_entry.proto [hunk 18]
```diff
@@ -75,28 +75,112 @@ import "networking/v1alpha3/gateway.proto";
...
+//   - number: 8888
+//     name: tcp-vm
+//     protocol: tcp
+//   endpoints:
+//     - addresses:
+//       - 1.2.3.4,1.2.3.5
+// ```
+//
+// The following example declares a workloadEntry running with DNS name httpbin.testcompany.com
+// and expose ports on 27018 and 8888
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]:
```suggestion
// which exposes ports 27018 and 8888.
```

File: networking/v1alpha3/workload_entry.proto [hunk 19]
```diff
@@ -75,28 +75,112 @@ import "networking/v1alpha3/gateway.proto";
...
+// +genclient
+// +k8s:deepcopy-gen=true
+// -->
 
 message WorkloadEntry {
 
-  string serviceAccountName = 1 [(google.api.field_behavior) = REQUIRED];
-  // One or more labels associated with the workloadEntry.
-  map<string, string> labels = 2;
+  // Identity information for this workloadEntry
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]: How about:

The service account that this WorkloadEntry is running as. Services calling this WorkloadEntry will verify that the identity matches the service account specified here.

File: networking/v1alpha3/workload_entry.proto [hunk 20]
```diff
@@ -75,28 +75,112 @@ import "networking/v1alpha3/gateway.proto";
...
+  string service_account_name = 1 [(google.api.field_behavior) = REQUIRED];
 
   // The ports associated with the workloadEntry.
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]: We should describe what these are used for.

File: networking/v1alpha3/workload_entry.proto [hunk 21]
```diff
@@ -75,28 +75,112 @@ import "networking/v1alpha3/gateway.proto";
...
   repeated Port ports = 3;
 
-  // One or more endpoints associated with the workloadEntry.
+  // One or more endpoints associated with the service.
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]: We should say something about how these are used too. These are the endpoints that traffic will be routed to for Services that select this WorkloadEntry.

File: networking/v1alpha3/workload_entry.proto [hunk 22]
```diff
@@ -75,28 +75,112 @@ import "networking/v1alpha3/gateway.proto";
...
   repeated Endpoint endpoints = 4 [(google.api.field_behavior) = REQUIRED];
 
   };
 
   message Endpoint {
 
+    // The endpoint addresses associated with the workloadEntry. Could be a DNS
+    // name as well as IP address
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]: Mention that multiple addresses are possible for things like dual stack.

File: networking/v1alpha3/workload_entry.proto [hunk 23]
```diff
@@ -75,28 +75,112 @@ import "networking/v1alpha3/gateway.proto";
...
     repeated string addresses = 1 [(google.api.field_behavior) = REQUIRED];
 
+    // topology contains arbitrary topology information associated with the
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]:
```suggestion
    // Topology information associated with the
```

File: networking/v1alpha3/workload_entry.proto [hunk 24]
```diff
@@ -75,28 +75,112 @@ import "networking/v1alpha3/gateway.proto";
...
+    // endpoint. These key/value pairs must conform with the label format.
```
Comment (smawson) [inline networking/v1alpha3/workload_entry.proto]: Rather than copy-pasting the K8s topology comment, maybe we can just point over at it? We can say this is identical to the topology information on an endpointslice that was derived from a Pod.

We can do something similar for the endpoints btw, point over to k8s endpointslice for additional details.

File: networking/v1alpha3/workload_entry.proto [hunk 25]
```diff
@@ -0,0 +1,186 @@
...
+//   name: myvm
+//   labels:
+//      app: myapp
+//      version: v1
+// spec:
+//   serviceAccountName: vmtest
+//   ports:
+//   - number: 27018
+//     name: mongodb
+//     protocol: MONGO
```
Comment (rshriram) [inline networking/v1alpha3/workload_entry.proto]: why does this have ports with protocols? How is this going to be used in Pilot?

File: networking/v1alpha3/workload_entry.proto [hunk 26]
```diff
@@ -0,0 +1,186 @@
...
+//   - number: 8888
+//     name: tcp-vm
+//     protocol: tcp
+//   endpoints:
+//     - addresses:
+//       - 1.2.3.4,1.2.3.5
```
Comment (rshriram) [inline networking/v1alpha3/workload_entry.proto]: Is this some comma separated string? What is a dual IP stack? If you want to declare one endpoint with multiple IPs, then this is not the place. Nothing in Istio is equipped to handle the dual IP in an endpoint (all the filter chains we setup in inbound, the iptables loop prevention, etc. will break down. There is nothing in place to prevent one IP from sending traffic to another and avoid traffic loop)

File: networking/v1alpha3/workload_entry.proto [hunk 27]
```diff
@@ -0,0 +1,186 @@
...
+// +kubetype-gen
+// +kubetype-gen:groupVersion=networking.istio.io/v1alpha3
+// +genclient
+// +k8s:deepcopy-gen=true
+// -->
+
+message WorkloadEntry {
+
+  // Identity information for this workloadEntry
+  string service_account_name = 1 [(google.api.field_behavior) = REQUIRED];
```
Comment (rshriram) [inline networking/v1alpha3/workload_entry.proto]: Why is this required? What happens if the end user is not using mtls to this workload as its a legacy VM? And secondly, what is the significance of this service account as far as Istio is concerned? How does this interact with the service entry's SAN field?

File: networking/v1alpha3/workload_entry.proto (hunk 22 above)
Comment (rshriram) [reply networking/v1alpha3/workload_entry.proto]: we can't handle dual stack today. A whole lot of system is designed based on the assumption that there is only one IP (specifically config generated by Pilot)

File: networking/v1alpha3/workload_entry.proto (hunk 24 above)
Comment (rshriram) [reply networking/v1alpha3/workload_entry.proto]: btw, we have our own notion of locality. Let us please not create yet another topology alternative

Comment (louiscryan) [reply networking/v1alpha3/workload_entry.proto]: What alignment can we achieve with

https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#topology

can we say that any of our labels are synonyms with those above ?

@howardjohn

**Comment by irisdingbj** [comment]
@rshriram Thanks for taking care of this. Noticed you have made it work so I will close this.

**Comment by howardjohn** [comment]
Thanks for getting this started Iris!
